### ISSUE TITLE
Bug: v6 `typeMatchesSpecifier` `FileSpecifier` checks in node modules

### ISSUE THREAD

**Issue by RebeccaStevens** [typescript-eslint/typescript-eslint#6819]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Repro Code

```typescript
typeMatchesSpecifier(
  type,
  {
    from: "file",
    name: "RegExp",
  },
  parserServices.program)
```

### Expected Result

It shouldn't find this type (as I haven't defined it).

### Actual Result

I finds it in "node_modules/.pnpm/typescript@5.0.3/node_modules/typescript/lib/lib.es5.d.ts".

### Additional Info

I would expect a `FileSpecifier` to ignore node module files by default.
`typescript/lib/*.d.ts` files should only match with a `LibSpecifier` and other files in node module to only match with a `PackageSpecifier`.

Maybe `FileSpecifier` needs an extra property like `ignorePath` or `ignoreGlob` that defaults to `"node_modules"`.
Or alternatively, `path` could be made a glob that defaults to selecting files not in `"node_modules"`.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `6.0.0-alpha.106` |
| `@typescript-eslint/parser`            | `6.0.0-alpha.106` |
| `@typescript-eslint/scope-manager`     | `6.0.0-alpha.106` |
| `@typescript-eslint/typescript-estree` | `6.0.0-alpha.106` |
| `@typescript-eslint/type-utils`        | `6.0.0-alpha.106` |
| `@typescript-eslint/utils`             | `6.0.0-alpha.106` |
| `TypeScript`                           | `5.0.3` |
| `ESLint`                               | `8.36.0` |
| `node`                                 | `18.15.0` |

**Comment by JoshuaKGoldberg** [comment]
🤔 @RebeccaStevens I think this makes sense. But similar to #6818 - is there a more full repro you can provide?

**Comment by RebeccaStevens** [comment]
I made a repo here: https://github.com/RebeccaStevens/typescript-eslint-issue-6818
The invalid test is failing.

The test rule I setup simply reports if `typeMatchesSpecifier` returns `false` for a parameter type.

I also found another strange issue when setting this up. I tried using `Set` for this test instead of `RegExp`, but the `Set` type would only find the default lib Set, and not the custom one.

**Comment by RebeccaStevens** [comment]
@marekdedic

**Comment by marekdedic** [comment]
Hi,
yeah, this makes sense.

I propose to think about the behaviour in the two situations separately:

1) The `FileSpecifier` doesn't contain a `path`. In this case, I think it makes sense not to match anything in `node_modules`

2) It does - well, I can go either way. I mean, you really **should** be using `PackageSpecifier` so I am OK with disallowing `node_modules`, but if you explicitly specify a path in `node_modules`, well, you seem to be misusing this intentionally...

Anyway, I can make a PR if there is some concensus on how this actually *should* work - it shouldn't be too hard to fix...

**Comment by RebeccaStevens** [comment]
@marekdedic Just beat you to it 😛

**Comment by RebeccaStevens** [comment]
I should probably open a new issue for this, but should we add the ability to specify a partial path?

So for example:
```json5
{ from: "file", name: "Foo", path: "src" }
```
Will match any `Foo` where there exists a type definition for it in a file that starts with `program.getCurrentDirectory().toLowerCase() + "src"`.

This will act similarly to how when no `path` is given, the file must start with `program.getCurrentDirectory().toLowerCase()`.

**Comment by JoshuaKGoldberg** [comment]
> I should probably open a new issue for this

Yes please 😇

**Comment by marekdedic** [comment]
Well or maybe this should be a part of #6819 ?
